src/nm-checkpoint.c

```diff
@@ -615,14 +691,17 @@ nm_checkpoint_new (NMManager *manager, GPtrArray *devices, guint32 rollback_time
 	for (i = 0; i < devices->len; i++) {
 		NMDevice *device = devices->pdata[i];
 
-		/* As long as the check point instance exists, it will keep a reference
-		 * to the device also if the device gets removed (by rmmod or by deleting
-		 * a connection profile for a software device). */
+		nm_assert (nm_dbus_object_get_path (NM_DBUS_OBJECT (device)));
 		g_hash_table_insert (priv->devices,
-		                     device,
+		                     g_strdup (nm_dbus_object_get_path (NM_DBUS_OBJECT (device))),
```

Comment:
as said above, it's wrong to stick to a D-Bus path after the device gets unexpected. As such, it makes no sense to clone and preserve this path.

You can assume that a D-Bus path is never reused. Hence, you can instead just index the hash by `g_direct_hash()` and the "device".

If you ever want to find a `NMDBusObject` by path, then use `nm_dbus_manager_lookup_object()`. That is a hash lookup by path of all currently exported objects.


You need to keep track in `priv->devices` of devices which are still exported, and those that were but no longer are. For that, you don't need a new artificial key (the path) that preserves longer. Instead:

```
static void
_device_removed (NMManager *manager, NMDevice *device, gpointer user_data)
{
    NMCheckpoint *self = NM_CHECKPOINT (user_data);
    NMCheckpointPrivate *priv = NM_CHECKPOINT_GET_PRIVATE (self);
    DeviceCheckpoint *dev_checkpoint;

    dev_checkpoint = g_hash_table_lookup (priv->devices, device);
    if (!dev_checkpoint)
        return;
    g_hash_table_steal (priv->devices, device);
    nm_clear_g_signal_handler (dev_checkpoint->device,
                                    &dev_checkpoint->dev_exported_change_id);
    g_clear_object (&dev_checkpoint->device);
    
    // MOVE THE DELETED ENTRY TO AN ALTERNATIVE LIST
    if (!priv->deleted_objects)
          priv->deleted_objects = g_ptr_array_new_with_free_fun (...);
    g_ptr_array_add (priv->deleted_objects, dev_checkpoint);

    _notify (self, PROP_DEVICES);
}
```

Reply:
I agree this could be a viable alternative.
The reasoning beyond using the dbus path as the key, was to keep a 1:1 mapping with what the API received as arguments. It doesn't look worse nor better to me than the approach proposed here: just different.

Reply:
Updated branch as requested: keep NMDevice as the g_hash_table key and add a GPtrArray to track removed devices.